### PR TITLE
ID Card Refactor Part 1: The Trim

### PR DESCRIPTION
**Pull request by Timberpoes**
## Test Merge Candidate

This PR is currently lacking in robust testing. There have been a number of sweeping changes including repathing. There are no gameplay functionality changes, but I would highly suggest testmerging this to make sure I haven't introduced any major runtimes or bugs in my repathing and refactoring efforts, especially on the visual side. I don't got the time to test every single change, but I've tested out as many of the systems as I can (as is evidenced by the gifs below).

## About The Pull Request

Advancing the implementation of https://hackmd.io/@tgstation/HkS-z64NP

Separates the concept of "retro" ID cards that use single icon states from "advanced" ID cards that are a composite of overlays.

Advanced ID cards are a base template - For example, we have the orange prisoner cards, black deathsquad/syndicate ones, gold, silver, grey - Without any hardcoded icons or any other details.

The assigned text, the little overlay that happens when the ID card has a name assigned to it, has similarly been pulled out to an overlay in all cases. This allow for future customisation if necessary - Any card can have any of the 3 assign text templates overlayed.

The card's assignment is now colloquially known as its trim. Currently there's no functionality change, but dynamic card "trimming" sets the stage for the trim portion of the design doc.

As a result of the additional flexibility afforded, this does mean players can now have Syndicate, Death Squad. ERT and NT assignments on their ID cards. I expect players to abuse this for fun and/or profit until it's no longer cool, unique or funny.

X22mkYTNM6
Fso72RziTF

Massive re-pathing, although there should be no functionality changes in practice (beyond being able to customise formerly hardcoded cards like ERT cards, syndie/deathsquad black cards and prisoner cards)

Also to accomplish this I needed to atomise some icons.
image

I've also created a non-processing IDs and Access subsystem, which should eventually hold most ID card and access related stuff. It currently only holds a dictionary of "occupation" names (such as Assistant, Central Command, Syndicate Operative, etc) to icon_states. The intent is that this SS will end up like SSjob where it contains a number of useful procs and dictionaries to help manage ID cards and access in general. Porting existing global procs over to it is for another PR.

## Why It's Good For The Game

Advances towards the ID card design doc. More versatile and modular ID cards.

## Changelog
:cl:
refactor: Modified ID cards to be more flexible and accept a wider variety of job trims.
/:cl:

### CODE REVIEW

**Comment by Fikou** [comment]
captain icon still is dirty grey  and not white 0/10

**Review by Twaticus — APPROVED** [review]

File: code/controllers/subsystem/id_access.dm [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+/datum/controller/subsystem/id_access/Initialize(timeofday)
+	setup_trims()
+	return ..()
+
+/// Populates the trims_by_title dictionary.
+/datum/controller/subsystem/id_access/proc/setup_trims()
+	var/list/all_hud_jobs = get_all_job_icons() + get_all_prisoner_jobs()
+
+	for(var/title in all_hud_jobs)
+		trims_by_title[title] = "trim_[ckey(title)]"
```
Comment (Rohesie) [inline code/controllers/subsystem/id_access.dm:18]: interesting usage of the `ckey()` proc 🎉

File: code/modules/mining/machine_vending.dm [hunk 2]
```diff
@@ -213,7 +213,7 @@
 /obj/machinery/mineral/equipment_vendor/golem/Initialize()
 	desc += "\nIt seems a few selections have been added."
 	prize_list += list(
-		new /datum/data/mining_equipment("Extra Id",       				/obj/item/card/id/mining, 				                   		250),
+		new /datum/data/mining_equipment("Extra Id",       				/obj/item/card/id/advanced/mining,								250),
```
Comment (Rohesie) [inline code/modules/mining/machine_vending.dm:216]:
```suggestion
		new /datum/data/mining_equipment("Extra Id", /obj/item/card/id/advanced/mining, 250),
```
I swear, this pattern of adding a million spaces is diff hell.
Can you just replace these with a space each? Or just accept the suggestion if you don't want to go there.

**Review by Rohesie — CHANGES_REQUESTED** [review]
Great PR. Do the changes and this is ready for testing.

**Comment by Timberpoes** [comment]
There's a bigger, badder, more insane PR incoming.

**Pull request closed without merging**